fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubber.java

```diff
@@ -0,0 +1,359 @@
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            RecordCursor<FDBIndexedRecord<Message>> cursor =
+                    store.scanIndexRecords(index.getName(), IndexScanType.BY_VALUE, tupleRange, null, IndexOrphanBehavior.RETURN, scanProperties);
+
+            final AtomicBoolean hasMore = new AtomicBoolean(true);
+            final AtomicReference<RecordCursorResult<FDBIndexedRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final long quota = spolicy().getQuota();
+
+            return iterateRangeOnly(store, cursor,
+                    this::deleteIndexIfDangling,
+                    lastResult, hasMore, recordsScanned)
+                    .thenApply(vignore -> hasMore.get() ?
+                                          lastResult.get().get().getIndexEntry().getKey() :
+                                          rangeEnd)
+                    .thenCompose(cont -> rangeSet.insertRange(store.ensureContextActive(), packOrNull(rangeStart), packOrNull(cont), true)
+                            .thenApply(ignore -> {
+                                if ( quota > 0 ) {
+                                    scanCounterIndexes += recordsScanned.get();
+                                    if (quota <= scanCounterIndexes) {
+                                        return false;
+                                    }
+                                }
+                                return !Objects.equals(cont, rangeEnd);
+                            }));
+        });
+    }
+
+    @Nullable
+    private CompletableFuture<FDBStoredRecord<Message>> deleteIndexIfDangling(FDBRecordStore store, final RecordCursorResult<FDBIndexedRecord<Message>> cursorResult) {
+        // This will always return null (!) - but sometimes will delete a dangling index
+        FDBIndexedRecord<Message> indexResult = cursorResult.get();
+
+        if (! indexResult.hasStoredRecord() ) {
+            // Here: Oh, No! this index is dangling!
+            final FDBStoreTimer timer = getRunner().getTimer();
+            timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_SCRUBBER_INDEXES_DANGLING);
+            final IndexEntry indexEntry = indexResult.getIndexEntry();
+            final Tuple valueKey = indexEntry.getKey();
+            final byte[] keyBytes = store.indexSubspace(common.getIndex()).pack(valueKey);
+
+            if (LOGGER.isWarnEnabled() && spolicy().shouldReportDangling()) {
+                LOGGER.warn(KeyValueLogMessage.build("Scrubber: dangling index entry",
+                        LogMessageKeys.KEY, valueKey.toString())
+                        .addKeysAndValues(common.indexLogMessageKeyValues())
+                        .toString());
+            }
+            if (spolicy().allowRepair()) {
+                // remove this index
+                store.getContext().ensureActive().clear(keyBytes);
```

Comment:
I think this needs to add a conflict range on the (missing) record that it's deleting the index entry of (or not scan the index at `SNAPSHOT`), because I think this is susceptible to the following concurrency bug:

1. A dangling entry exists pointing to a record that doesn't exist
2. Transaction 1 scans the index and sees that it's missing a record, deletes the entry (but not yet committing)
3. Transaction 2 inserts a new record with that primary key and index value, overwriting the old, previously dangling entry and commits
4. Transaction 1 commits (successfully), deleting the new entry, which now means that the record is unindexed

Scanning the index at `SERIALIZABLE` would fix the problem, as operation 4 would mean the read of the index entry step 1 conflicts with step 3 inserting the new index entry. Adding an explicit read conflict range on the (missing) record would also fix the problem as that read conflict range would conflict with the record write in step 3.

It would also be good to get a test of this...somehow, though that test is somewhat tricky to write without a transactional scrubbing variant was created (that scrubs a whole index in one go). Then I think you could create a dangling index entry and start two transactions, one which inserted a replacement record and another that scrubbed the whole index, and make sure that the right operations failed.

Reply:
Oh, that answers my previous question. I'll look into adding an explicit read conflict range.

Reply:
(After consulting with Alec offline)
Added the index entry itself to the conflict list.